packages/xl-ai/src/prosemirror/changeset.ts

```diff
@@ -0,0 +1,340 @@
+
+  // if the docs are not equal, we might be missing changes that have not been detected by prosemirror-changeset,
+  // such as marks or node attributes. Let's find and add the missing changes.
+  //
+  // example, using _ for bold:
+  // actual:   helloworld!
+  // expected: hello_world_!
+  //
+  // actual:   hello_world_!
+  // expected: helloworld!
+  //
+  // note: might be possible to merge this loop with the one at the start of the function
+  // but considered ok for now
+  while (diffStart !== null) {
+    const expectedNode = expectedDoc.resolve(diffStart).nodeAfter;
+    const actualNode = tr.doc.resolve(diffStart).nodeAfter;
+
+    if (!expectedNode || !actualNode) {
+      throw new Error("diffNode not found");
+    }
+
+    const isNodeAttrChange =
+      !expectedNode.isLeaf && expectedNode.content.eq(actualNode.content);
+
+    const length = isNodeAttrChange
+      ? 1
+      : Math.min(expectedNode.nodeSize, actualNode.nodeSize);
+
+    const to = diffStart + length;
+    const fromA = invertMap.map(diffStart);
+    const toA = invertMap.map(to);
+
+    // find the position in changes array to insert the new change
+    let insertPos = changes.length;
+    for (let i = 0; i < changes.length; i++) {
+      if (changes[i].fromA >= toA) {
+        insertPos = i;
+        break;
+      }
+    }
+
+    changes.splice(insertPos, 0, {
+      fromA: fromA,
+      toA: toA,
+      fromB: diffStart,
+      toB: to,
+      deleted: [],
+      inserted: [],
+      type: isNodeAttrChange ? "node-type-or-attr-update" : "mark-update",
+    });
+
+    // apply the step so we can find the next diff
+    // Note: even for node type changes, this works - although maybe a ReplaceAroundStep might be cleaner?
+    tr.step(
+      new ReplaceStep(
+        diffStart,
+        to,
+        expectedDoc.slice(diffStart, to),
+        isNodeAttrChange,
+      ),
+    );
+    const newDiffStart = tr.doc.content.findDiffStart(expectedDoc.content);
+
+    if (newDiffStart === diffStart) {
+      // prevent infinite loop, should not happen
+      throw new Error("diffStart not moving");
+    }
+
+    diffStart = newDiffStart;
+  }
+
+  return changes;
+}
+
+const createEncoder = (doc: Node, updatedDoc: Node) => {
+  // this encoder makes sure unchanged table cells stay intact,
+  // without this, prosemirror-changeset would too eagerly
+  // return changes across table cells (this is covered in test cases).
+
+  const tableCellsOld = new Set<string>();
+  const tableCellsNew = new Set<string>();
+  doc.descendants((node) => {
+    if (node.type.name === "tableCell") {
+      tableCellsOld.add(JSON.stringify(node.toJSON()));
+    }
+  });
+
+  updatedDoc.descendants((node) => {
+    if (node.type.name === "tableCell") {
+      tableCellsNew.add(JSON.stringify(node.toJSON()));
+    }
+  });
+
+  const tableCells = new Set(
+    [...tableCellsOld].filter((cell) => tableCellsNew.has(cell)),
+  );
```

Comment:
const tableCells = tableCellsOld.intersection(tableCellsNew);

Reply:
do you know if this consistently works or requires extra setup? I tried this but got an error at some point that `intersection` is not available

Reply:
Hm, it's a newer feature, but should be reasonably available: https://caniuse.com/?search=Set%20intersection

No, biggie yours works too